spectrum-network/src/protocol_upgrade.rs

```diff
@@ -33,6 +33,8 @@ pub enum ProtocolHandshakeErr {
 pub enum ProtocolUpgradeErr {
     #[error(transparent)]
     HandshakeErr(#[from] ProtocolHandshakeErr),
+    #[error("Unsupported {0:?}")]
+    UnsupportedProtocolVer(ProtocolVer),
```

Comment:
I looked into it in more detail; I see what you mean. Thanks for the `Parse, don't validate` article, learnt a lot. Do you think it's worth wrapping up these collections in a newtype which prevents any mutation of the underlying collection, and also using a newtype-wrapped protocol id/ver?

e.g. in `ProtocolUpgradeIn::supported_versions`, we wrap the `BTreeMap` in a `struct SupportedProtocolVers`,  and instead of handling `ProtocolVer` directly we have `struct SupportedProtocolVer(ProtocolVer)` which can only be generated by `SupportedProtocolVers`?

If that's too much, we can comment the use of the `unwrap` and explicitly allow it via `#[allow(clippy::unwrap_used]`. Later when we setup CI/CD, we can add clippy exclusions like e.g. https://github.com/ergoplatform/sigma-rust/blob/develop/ergotree-ir/src/lib.rs#L13-L19

Reply:
Yes, wrappers that hold some assumption should be good.

Reply:
Implemented the wrappers.